Don't hide tooltip when pointer is over it.

grokys opened the pull request:
## What does the pull request do?

As described in #8638, don't hide the tooltip for a control when the pointer moves over it.

## What is the current behavior?

If a tooltip should be shown near the bounds of the screen, it won't appear. You can test this on the tooltip page in the control catalog.

## What is the updated/expected behavior with this PR?

The tooltip stays open.

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #8638

grokys committed:
Add failing test for #8638.

grokys committed:
Don't hide tooltip when pointer is over tooltip.

Fixes #8638

MrJul commented:
Shouldn't the tooltip close when the mouse isn't over it anymore?
With this PR, the user has to move the mouse back from the tooltip to the original control, then out of it again for the tooltip to disappear.

ltetak commented:
I think this should solve this https://github.com/AvaloniaUI/Avalonia/pull/13709

grokys commented:
> Shouldn't the tooltip close when the mouse isn't over it anymore?
> With this PR, the user has to move the mouse back from the tooltip to the original control, then out of it again for the tooltip to disappear.

Good point. I guess if we go with this approach we need to check for the pointer exiting the tooltip and then close if it's over another control.

@MrJul not sure what you think of this approach vs #13709 ? I suspect that capturing the mouse to show a tooltip is going to break things.

MrJul commented:
> Good point. I guess if we go with this approach we need to check for the pointer exiting the tooltip and then close if it's over another control.

Yes, I think that would solve the issue.

> @MrJul not sure what you think of this approach vs #13709 ? I suspect that capturing the mouse to show a tooltip is going to break things.

I've commented on #13709: I agree with you, the capture approach is breaking things.

grokys committed:
Close the tooltip when pointer exits.

If the pointer has been moved from the control to the tooltip, then out of the tooltip to another control, ensure that the tooltip is closed.

grokys commented:
@MrJul ok, pushed a fix for the pointer exiting the tooltip.

MrJul commented on src/Avalonia.Controls/ToolTip.cs:
```diff
@@ -61,6 +61,12 @@ public class ToolTip : ContentControl, IPopupHostProvider
+        /// <summary>
+        /// Stores the control that a <see cref="ToolTip"/> is being shown on.
+        /// </summary>
+        internal static readonly AttachedProperty<Control?> ToolTipAdornedControlProperty =
```
Can probably be simplified to a standard CLR property since the owner is the `ToolTip` itself and it's internal.

grokys replied:
Yeah that would make sense.

grokys replied:
Done.

MrJul commented:
> @MrJul ok, pushed a fix for the pointer exiting the tooltip.

I've tested the PR and it reacts exactly as I would expect :)

grokys committed:
AdornedControl can be a standard CLR property.

MrJul reviewed (approved):
LGTM!

grokys committed:
Merge branch 'master' into fixes/8638-pointer-over-tooltip

ltetak commented:
Sorry for commenting on an old thread. This solution works great on Windows but does not work on macOS and probably does not work on Linux as well (https://github.com/AvaloniaUI/Avalonia/issues/11179). 
IsPointerOver is not set quickly enough and the tooltip does not have IsPointerOver set yet when the ControlPointerExited is invoked.
The problem is fundamentally in the order of two events: `RawPointerEventType.LeaveWindow `(for leaving the main window) and other events that activate the ToolTip popup window. The window is not activated immediately on mac. One needs to move the cursor above the window to get it activated. This delay causes the tooltip to be closed immediately.
I believe the solution would be to find some native event that would notify Avalonia layer ASAP when the window has been created and the mouse is over it without waiting for the mouse movement.
@grokys
